File: products/compliance/src/analyzers/dep-auditor.ts
import type { Analyzer, Finding, RuleDefinition, ScannedFile } from '../types.js';
import { readFileSync, existsSync } from 'node:fs';
import { dirname, join, basename } from 'node:path';

/**
 * Audits project dependencies for compliance issues.
 *
 * Static checks (no external commands):
 * - Unpinned production dependencies (^ or ~ version ranges)
 * - Missing lockfile (package-lock.json or npm-shrinkwrap.json)
 * - Missing engines field in package.json
 * - Copyleft license (GPL, AGPL, LGPL) in package.json
 */
export class DepAuditor implements Analyzer {
  name = 'dep-auditor';

  async analyze(files: ScannedFile[], rules: RuleDefinition[]): Promise<Finding[]> {
    const myRules = rules.filter((r) => r.analyzer === this.name);
    if (myRules.length === 0) return [];

    // Build lookup from pattern -> rules (multiple rules may share a pattern across frameworks)
    const rulesByPattern = new Map<string, RuleDefinition[]>();
    for (const rule of myRules) {
      const existing = rulesByPattern.get(rule.pattern) ?? [];
      existing.push(rule);
      rulesByPattern.set(rule.pattern, existing);
    }

    const findings: Finding[] = [];

    const packageFiles = files.filter((f) => basename(f.path) === 'package.json');

    for (const pkgFile of packageFiles) {
      let pkg: Record<string, unknown>;
      try {
        const content = readFileSync(pkgFile.path, 'utf-8');
        pkg = JSON.parse(content) as Record<string, unknown>;
      } catch {
        continue;
      }

      // ── Unpinned dependencies ────────────────────────────────────
      this.checkUnpinnedDeps(pkg, pkgFile, rulesByPattern, findings);

      // ── Missing lockfile ─────────────────────────────────────────
      this.checkMissingLockfile(pkgFile, rulesByPattern, findings);

      // ── Missing engines ──────────────────────────────────────────
      this.checkMissingEngines(pkg, pkgFile, rulesByPattern, findings);

      // ── Copyleft license ─────────────────────────────────────────
      this.checkCopyleftLicense(pkg, pkgFile, rulesByPattern, findings);
    }

    return findings;
  }

  /**
   * Check production dependencies for unpinned version ranges (^ or ~).
   * Creates ONE finding per package.json listing all unpinned deps in the evidence.
   */
  private checkUnpinnedDeps(
    pkg: Record<string, unknown>,
    pkgFile: ScannedFile,
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const deps = pkg.dependencies as Record<string, string> | undefined;
    if (!deps) return;

    const unpinned = Object.entries(deps).filter(
      ([, version]) => version.startsWith('^') || version.startsWith('~'),
    );

    if (unpinned.length > 0) {
      this.addFindings(findings, 'unpinned-deps', rulesByPattern, {
        file: pkgFile.relativePath,
        evidence: `Unpinned: ${unpinned.map(([name]) => name).join(', ')}`,
      });
    }
  }

  /**
   * Check if a lockfile (package-lock.json or npm-shrinkwrap.json) exists
   * in the same directory as the package.json.
   */
  private checkMissingLockfile(
    pkgFile: ScannedFile,
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const dir = dirname(pkgFile.path);
    const hasLockfile =
      existsSync(join(dir, 'package-lock.json')) ||
      existsSync(join(dir, 'npm-shrinkwrap.json'));

    if (!hasLockfile) {
      this.addFindings(findings, 'missing-lockfile', rulesByPattern, {
        file: pkgFile.relativePath,
        evidence: 'No package-lock.json or npm-shrinkwrap.json found',
      });
    }
  }

  /**
   * Check if package.json has an engines field.
   */
  private checkMissingEngines(
    pkg: Record<string, unknown>,
    pkgFile: ScannedFile,
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    if (!pkg.engines) {
      this.addFindings(findings, 'missing-engines', rulesByPattern, {
        file: pkgFile.relativePath,
        evidence: 'No engines field in package.json',
      });
    }
  }

  /**
   * Check if the license field in package.json indicates a copyleft license
   * (GPL, AGPL, LGPL).
   */
  private checkCopyleftLicense(
    pkg: Record<string, unknown>,
    pkgFile: ScannedFile,
    rulesByPattern: Map<string, RuleDefinition[]>,
    findings: Finding[],
  ): void {
    const license = pkg.license as string | undefined;
    if (!license) return;

    const upper = license.toUpperCase();
    if (upper.includes('GPL') || upper.includes('AGPL') || upper.includes('LGPL')) {
      this.addFindings(findings, 'copyleft-license', rulesByPattern, {
        file: pkgFile.relativePath,
        evidence: `Copyleft license detected: ${license}`,
      });
    }
  }

  /**
   * Create findings for all rules matching a given pattern.
   * A single pattern may map to rules in multiple frameworks (soc2, hipaa, gdpr).
   */
  private addFindings(
    findings: Finding[],
    pattern: string,
    rulesByPattern: Map<string, RuleDefinition[]>,
    context: { file?: string; line?: number; evidence?: string },
  ): void {
    const matchedRules = rulesByPattern.get(pattern);
    if (!matchedRules || matchedRules.length === 0) return;

    for (const rule of matchedRules) {
      findings.push({
        id: rule.id,
        title: rule.title,
        description: rule.description,
        severity: rule.severity,
        framework: rule.framework,
        controlIds: rule.controls,
        file: context.file,
        line: context.line,
        evidence: context.evidence,
        analyzer: this.name,
        fixable: rule.fixable,
      });
    }
  }
}

File: products/compliance-go/analyzers/dep_auditor.go
package analyzers

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/rishav1305/soul/products/compliance-go/rules"
)

// DepAuditor implements the Analyzer interface for auditing dependency
// configurations: unpinned versions, missing lockfiles, missing engines
// field, and copyleft license detection.
type DepAuditor struct{}

// Name returns the analyzer identifier used to match rules.
func (d *DepAuditor) Name() string {
	return "dep-auditor"
}

// packageJSON represents the relevant fields of a package.json file.
type packageJSON struct {
	Name            string            `json:"name"`
	Engines         json.RawMessage   `json:"engines"`
	Dependencies    map[string]string `json:"dependencies"`
	DevDependencies map[string]string `json:"devDependencies"`
	License         string            `json:"license"`
}

// Analyze inspects package.json files and lockfile presence.
func (d *DepAuditor) Analyze(files []ScannedFile, allRules []rules.Rule) ([]Finding, error) {
	myRules := filterRules(allRules, d.Name())
	if len(myRules) == 0 {
		return nil, nil
	}

	rulesByPattern := groupRulesByPattern(myRules)
	var findings []Finding

	// Build path index
	pathSet := make(map[string]ScannedFile)
	for _, f := range files {
		pathSet[f.RelativePath] = f
	}

	// Find all package.json files
	var packageFiles []ScannedFile
	for _, f := range files {
		if filepath.Base(f.RelativePath) == "package.json" {
			packageFiles = append(packageFiles, f)
		}
	}

	for _, pkgFile := range packageFiles {
		findings = append(findings, d.analyzePackageJSON(pkgFile, files, rulesByPattern)...)
	}

	// If no package.json at all, skip lockfile/engines checks
	if len(packageFiles) == 0 {
		return findings, nil
	}

	// Check for lockfile presence (project-level)
	findings = append(findings, d.checkLockfile(files, rulesByPattern)...)

	return findings, nil
}

// analyzePackageJSON parses and inspects a single package.json file.
func (d *DepAuditor) analyzePackageJSON(pkgFile ScannedFile, files []ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	content, err := os.ReadFile(pkgFile.Path)
	if err != nil {
		return nil
	}

	var pkg packageJSON
	if err := json.Unmarshal(content, &pkg); err != nil {
		return nil
	}

	// Check for unpinned dependencies
	findings = append(findings, d.checkUnpinnedDeps(pkgFile, pkg.Dependencies, "dependencies", rulesByPattern)...)
	findings = append(findings, d.checkUnpinnedDeps(pkgFile, pkg.DevDependencies, "devDependencies", rulesByPattern)...)

	// Check for missing engines field
	if len(pkg.Engines) == 0 || string(pkg.Engines) == "null" {
		for _, rule := range rulesByPattern["missing-engines"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				File:        pkgFile.RelativePath,
				Evidence:    "package.json is missing the engines field",
				Analyzer:    d.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	// Check for copyleft license in the package itself
	if isCopyleftLicense(pkg.License) {
		for _, rule := range rulesByPattern["copyleft-license"] {
			findings = append(findings, Finding{
				ID:          rule.ID,
				Title:       rule.Title,
				Description: rule.Description,
				Severity:    rule.Severity,
				Framework:   rule.Framework,
				ControlIDs:  rule.Controls,
				File:        pkgFile.RelativePath,
				Evidence:    fmt.Sprintf("Package uses copyleft license: %s", pkg.License),
				Analyzer:    d.Name(),
				Fixable:     rule.Fixable,
			})
		}
	}

	return findings
}

// checkUnpinnedDeps checks for dependencies using ^ ~ or * version prefixes.
func (d *DepAuditor) checkUnpinnedDeps(pkgFile ScannedFile, deps map[string]string, section string, rulesByPattern map[string][]rules.Rule) []Finding {
	var findings []Finding

	for name, version := range deps {
		if isUnpinnedVersion(version) {
			for _, rule := range rulesByPattern["unpinned-deps"] {
				findings = append(findings, Finding{
					ID:          rule.ID,
					Title:       rule.Title,
					Description: fmt.Sprintf("%s (%s.%s: %s)", rule.Description, section, name, version),
					Severity:    rule.Severity,
					Framework:   rule.Framework,
					ControlIDs:  rule.Controls,
					File:        pkgFile.RelativePath,
					Evidence:    fmt.Sprintf("%s: %s@%s", section, name, version),
					Analyzer:    d.Name(),
					Fixable:     rule.Fixable,
				})
			}
		}
	}

	return findings
}

// checkLockfile checks for the presence of a lockfile in the project.
func (d *DepAuditor) checkLockfile(files []ScannedFile, rulesByPattern map[string][]rules.Rule) []Finding {
	lockfileNames := map[string]bool{
		"package-lock.json": true,
		"yarn.lock":         true,
		"pnpm-lock.yaml":    true,
	}

	for _, f := range files {
		base := filepath.Base(f.RelativePath)
		if lockfileNames[base] {
			return nil
		}
	}

	var findings []Finding
	for _, rule := range rulesByPattern["missing-lockfile"] {
		findings = append(findings, Finding{
			ID:          rule.ID,
			Title:       rule.Title,
			Description: rule.Description,
			Severity:    rule.Severity,
			Framework:   rule.Framework,
			ControlIDs:  rule.Controls,
			Evidence:    "No lockfile (package-lock.json, yarn.lock, pnpm-lock.yaml) found",
			Analyzer:    d.Name(),
			Fixable:     rule.Fixable,
		})
	}
	return findings
}

// isUnpinnedVersion returns true if the version string uses a range prefix.
func isUnpinnedVersion(version string) bool {
	v := strings.TrimSpace(version)
	if v == "" {
		return false
	}
	return strings.HasPrefix(v, "^") || strings.HasPrefix(v, "~") || v == "*"
}

// isCopyleftLicense checks if a license string indicates a copyleft license.
func isCopyleftLicense(license string) bool {
	upper := strings.ToUpper(strings.TrimSpace(license))
	copyleftKeywords := []string{"GPL", "AGPL", "LGPL", "EUPL", "MPL", "SSPL", "OSL"}
	for _, kw := range copyleftKeywords {
		if strings.Contains(upper, kw) {
			return true
		}
	}
	return false
}
